File: ando_library/src/main/java/com/ando/library/base/BaseFragment.java
package com.ando.library.base;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import com.ando.toolkit.StringUtils;
import com.ando.toolkit.ext.ToastUtils;

/**
 * Title:BaseFragment
 * <p>
 * Description:
 * </p>
 *
 * @author javakam
 * @date 2019/3/17 13:24
 */
public abstract class BaseFragment extends Fragment implements IBaseInterface {

    protected BaseActivity activity;
    protected View rootView;
    //
    protected boolean isActivityCreated;    //Activity 是否已创建
    protected boolean isVisibleToUser;      //Fragment 是否对用户可见
    protected boolean isHidden;               //hidden

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        activity = (BaseActivity) context;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (getLayoutId() > 0) {
            rootView = inflater.inflate(getLayoutId(), container, false);
        } else {
            rootView = getLayoutView();
        }

        initView(savedInstanceState);
        initListener();
        return rootView;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        isActivityCreated = true;
        initData();
    }

    /**
     * @deprecated Use {@link FragmentTransaction#setMaxLifecycle}
     */
    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        this.isVisibleToUser = isVisibleToUser;
        //L.i("setUserVisibleHint isVisibleToUser ===> " + isVisibleToUser);
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        this.isHidden = hidden;
        //L.i("onHiddenChanged hidden ===> " + hidden);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        rootView = null;
        activity = null;
    }

    /**
     * 通过Class跳转界面
     **/
    public void startActivity(Class<?> cls) {
        startActivity(cls, null);
    }

    /**
     * 通过Class跳转界面
     **/
    public void startActivityForResult(Class<?> cls, int requestCode) {
        startActivityForResult(cls, null, requestCode);
    }

    /**
     * 含有Bundle通过Class跳转界面
     **/
    public void startActivityForResult(Class<?> cls, Bundle bundle, int requestCode) {
        Intent intent = new Intent();
        intent.setClass(activity, cls);
        if (bundle != null) {
            intent.putExtras(bundle);
        }
        startActivityForResult(intent, requestCode);
    }

    /**
     * 含有Bundle通过Class跳转界面
     **/
    public void startActivity(Class<?> cls, Bundle bundle) {
        if (getActivity() != null) {
            Intent intent = new Intent();
            intent.setClass(activity, cls);
            if (bundle != null) {
                intent.putExtras(bundle);
            }
            startActivity(intent);
        }
    }

    public void shortToast(@StringRes int text) {
        ToastUtils.shortToast(StringUtils.noNull(getString(text)));
    }

    public void shortToast(String text) {
        ToastUtils.shortToast(StringUtils.noNull(text));
    }

    public void longToast(@StringRes int text) {
        ToastUtils.longToast(StringUtils.noNull(getString(text)));
    }

    public void longToast(String text) {
        ToastUtils.longToast(StringUtils.noNull(text));
    }

    public void controlBackPress() {
        //https://stackoverflow.com/questions/22552958/handling-back-press-when-using-fragments-in-android
        getView().setFocusableInTouchMode(true);
        getView().requestFocus();
        getView().setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                if (event.getAction() == KeyEvent.ACTION_UP && keyCode == KeyEvent.KEYCODE_BACK) {
                    // handle back button
                    // Toast.makeText(activity, "handle back button", Toast.LENGTH_SHORT).show();
                    onBackPressed();
                    return true;
                }
                return false;
            }
        });
    }

    public void onBackPressed() {
    }

}

File: ando_library/src/main/java/com/ando/library/base/BaseFragment.kt
package com.ando.library.base

import android.content.Context
import android.os.Bundle
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import com.ando.toolkit.ext.otherwise
import com.ando.toolkit.ext.yes

/**
 * Title:BaseFragment
 *
 * Description:
 *
 * @author javakam
 * @date 2019/3/17 13:24
 */
abstract class BaseFragment : Fragment() {

    //注:不要命名为 activity 会和 Fragment.getActivity 冲突
    protected lateinit var baseActivity: BaseActivity
        private set

    //
    var isActivityCreated = false//Activity 是否已创建
    var isVisibleToUser = false //Fragment 是否对用户可见
    var isHiddenToUser = false //hidden

    override fun onAttach(context: Context) {
        super.onAttach(context)
        baseActivity = context as BaseActivity
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        isActivityCreated = true
    }

    @Deprecated("Use {@link FragmentTransaction#setMaxLifecycle}")
    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        this.isVisibleToUser = isVisibleToUser
        //L.i("setUserVisibleHint isVisibleToUser ===> " + isVisibleToUser);
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        isHiddenToUser = hidden
        //L.i("onHiddenChanged hidden ===> " + hidden);
    }

    //https://stackoverflow.com/questions/22552958/handling-back-press-when-using-fragments-in-android
    fun controlBackPress() =
        view?.let { it ->
            it.isFocusableInTouchMode = true
            it.requestFocus()
            it.setOnKeyListener(View.OnKeyListener { v, keyCode, event ->
                if (event.action == KeyEvent.ACTION_UP && keyCode == KeyEvent.KEYCODE_BACK) {
                    // handle back button
                    // Toast.makeText(activity, "handle back button", Toast.LENGTH_SHORT).show();
                    onBackPressed()
                    return@OnKeyListener true
                }
                false
            })
        }

    fun onBackPressed() {}

}

abstract class BaseMvcFragment : BaseFragment(), IBaseInterface {

    protected lateinit var rootView: View
        private set

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        rootView = (getLayoutId() > 0).yes {
            inflater.inflate(getLayoutId(), container, false)
        }.otherwise {
            getLayoutView()!!
        }
        initView(savedInstanceState)
        initListener()
        return rootView
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initData()
    }

}

abstract class BaseMvvmFragment<T : ViewDataBinding> : BaseFragment() {

    abstract val layoutId: Int
    lateinit var binding: T
    abstract fun initView(root: View, savedInstanceState: Bundle?)

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = DataBindingUtil.inflate(inflater, layoutId, container, false)
        binding.lifecycleOwner = this
        binding.executePendingBindings()
        initView(binding.root,savedInstanceState)
        return binding.root
    }

//    override fun onViewCreated(root: View, savedInstanceState: Bundle?) {
//        super.onViewCreated(root, savedInstanceState)
//        initView(root,savedInstanceState)
//    }


}